proj3_chess/chess_pieces.py

```diff
@@ -0,0 +1,130 @@
+from board import Board
+from player import Player
+
+
+class ChessPiece:
+    """Mother class for all chess pieces."""
+
+    piece_name = ""
+
+    def __init__(self, player: Player, board: Board, spot: tuple) -> None:
+        # TODO: raise error if spot is already taken
+        self.player = player
+        self.player.pieces.append(self)  # add the piece to the player
+        self.name = player.symbol(self.piece_name)
+        self.board = board
+        self.spot = spot
+        self.board.board_state[spot[0]][spot[1]] = self  # add the piece to the board
```

Comment:
Same as above, weird to edit another object's member in this constructor.

Reply:
If you want to enforce a piece being added to the board when it is created, you can add a `create_piece` method to the board object that takes the piece type and position and player and all that

Reply:
Isnt that essentially the meaning of a constructor though? Creating the obj?

Reply:
But you are also editing the inner state of a different object, which creates a deep coupling between the objects. Changing one of these classes will most likely break the other. We want to aim to minimize how dependent our classes are on each other.

The heart of this problem is that you have a double-reference to the piece's location, which must always be the same value - In the piece's `spot` property and in the board's `board_state` property. You have two possible solutions here:

* Get rid of one of these references - for example, you can claim that a piece doesn't have a position itself, but a board is a mapping of piece to position (for example, using a dictionary). Then the piece doesn't need a `spot` property, and can ask the board where it is.
* Find a better place to edit both of the references. A constructor, like you said, is meant to create an object, and thus it is not expected that it will alter another object's state. On the other hand, a method called `create_piece` in the `Board` object obviously creates a new piece and alters the board's state, from a place where it makes sense to do both.